tools/clang/lib/Sema/SemaHLSL.cpp

```diff
@@ -15790,41 +15888,71 @@ clang::FunctionDecl *ValidateNoRecursion(clang::Sema *self,
   // on functions that are unreachable (as an early form of dead code
   // elimination).
   if (FD) {
-    hlsl::CallGraphWithRecurseGuard CG;
-    CG.BuildForEntry(FD);
-    return CG.CheckRecursion(FD);
+    HLSLExternalSource *hlslSource = HLSLExternalSource::FromSema(self);
+    hlslSource->BuildCallGraphForEntry(FD);
+    return hlslSource->CheckForCallGraphRecursion(FD);
   }
   return nullptr;
 }
 
+// There is an assumption being made that this function will only be called
+// on library shader stages.Thus, there is no entry point, only exported
+// functions
 void ValidateNoRecursionInTranslationUnit(clang::Sema *self) {
   std::set<FunctionDecl *> FDecls;
   std::vector<FunctionDecl *> FDeclsVec;
+  std::vector<FunctionDecl *> AllExportedFDecls;
+
   for (auto decl : self->getASTContext().getTranslationUnitDecl()->decls()) {
-    // TODO: improve condition so that only exported functions are checked,
-    // instead of all functions. Issue: #5857
     if (FunctionDecl *FD = dyn_cast<FunctionDecl>(decl)) {
-      // returns the first recursive function declaration detected
-      // from this function declaration FD, and determines whether
-      // the recursion was detected in the patch-constant function
-      FunctionDecl *pResult = ValidateNoRecursion(self, FD);
-      // if there is a function that was detected to be recursive,
-      // then make sure it is saved for later to emit a diagnostic
-      if (pResult) {
-        FDecls.insert(pResult);
-        FDeclsVec.push_back(pResult);
+      if (!FD->hasBody() ||
+          !IsExported(self, FD, getDefaultLinkageExternal(self))) {
+        continue;
+      }
+      AllExportedFDecls.push_back(FD);
+    } else if (DeclContext *declContext = dyn_cast<DeclContext>(decl)) {
+      // Loop through all declarations inside the class or namespace
+      for (DeclContext::decl_iterator j = declContext->decls_begin(),
+                                      e = declContext->decls_end();
+           j != e; ++j) {
+        // Check if the declaration is a function declaration
+        if (FunctionDecl *functionDecl = dyn_cast<FunctionDecl>(*j)) {
+          if (!functionDecl->hasBody() ||
+              !IsExported(self, functionDecl,
+                          getDefaultLinkageExternal(self))) {
+            continue;
+          }
+          AllExportedFDecls.push_back(functionDecl);
```

Comment:
This gathers one level deep, but I think a recursive worklist pattern should be used to go deeper, and it's less code.

Basically, that's:
```cpp
  llvm::SmallVector<DeclContext *, 8> Worklist;
  Worklist.push_back(self->getASTContext().getTranslationUnitDecl());
  while (Worklist.size()) {
    DeclContext *DC = Worklist.pop_back_val();
    for (auto *D : DC->decls()) {
      if (auto *FD = dyn_cast<FunctionDecl>(D)) {
        if (FD->hasBody() &&
            IsExported(self, FD, getDefaultLinkageExternal(self)))
          AllExportedFDecls.push_back(FD);
      } else if (auto *DC2 = dyn_cast<DeclContext>(D)) {
        Worklist.push_back(DC2);
      }
    }
  }
```

Also, how is the main function checked for non-lib case?  I think this function could be split into a function that verifies recursion on a set of decls, and things that gather those decls we need to check.  Unification of code checking recursion and so on for all functions, reachability between hull shaders and patch constant functions, between library and non-library compilations is also possible, and desirable.  It's still a bit of a mess right now.

It appears to me that we have:
- if library
  - iterate all function decls, skipping no-body and internal linkage
    - if patch constant attr
      - look up patch constant function, error if not found
      - check patch constant function for recursion
      - check hull shader for recursion
      - check reachability between these
  - iterate all function decls, skipping no-body and internal linkage (again, but this time, go into decl context one level deep)
    - add function to AllExportedFDecls
  - check all functions in AllExportedFDecls for recursion (this will redundantly check hull shaders)
- else (not library)
  - lookup entry function
  - check for recursion
  - if hull shader
    - look up patch constant function, error if not found
    - check patch constant function for recursion
    - check reachability between these

This could be simplified to:
- gather function phase
  - if library
    - use recursive worklist pattern to gather functions
  - else
    - look up and add entry function
- check functions phase
  - check for recursion
  - if hull shader
    - get hull shader function, error if not found
    - check for recursion
    - check for reachability between these